clang/include/clang/3C/ConstraintsGraph.h

```diff
@@ -194,19 +200,20 @@ class DataGraph
   // is allocated. Node equality is defined only by the data stored in a node,
   // so if any node already contains the data, this node will be found.
   virtual NodeType *findOrCreateNode(Data D) {
-    auto *OldN = this->findNode(NodeType(D));
-    if (OldN != this->end())
-      return *OldN;
+    if (NodeSet.find(D) != NodeSet.end())
+      return NodeSet[D];
 
     auto *NewN = new NodeType(D);
-    this->addNode(*NewN);
+    this->Nodes.push_back(NewN);
+    NodeSet[D] = NewN;
```

Comment:
So `Nodes` is part of the `llvm::DirectedGraph` class, but `NodeSet` is something you are adding? Why is it needed?